### PR TITLE
#93  NHD via XML

### PR DESCRIPTION
**Pull request by llorllale**
As per #93 

From the description included in NHD.xsl:
> The NHD (Normalized Hamming Distance) class cohesion metric
          is based on the "Hamming distance" from information theory. Here,
          we measure the similarity in all methods of a class in terms of
          the types of their arguments. A class in which all methods accept
          the same set of types of parameters is said to be in "perfect
          parameter agreement" (NHD score "1"), whereas a class in which all
          methods accept unique parameter types not shared by others has
          no parameter agreement (NHD score "0").

### CODE REVIEW

**Comment by 0crat** [comment]
@yegor256[/z](https://www.0crat.com/u/yegor256) please, pay attention to this pull request

**Comment by 0crat** [comment]
Job [#113](https://github.com/yegor256/jpeek/issues/113) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#113](https://github.com/yegor256/jpeek/issues/113) is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil). The budget is 15 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§27](http://datum.zerocracy.com/pages/policy.html#27) and when you decide to accept the changes, inform @yegor256[/z](https://www.0crat.com/u/yegor256) (the architect) right in this ticket. If you decide that this PR should not be accepted ever, also inform the architect.

File: src/main/resources/org/jpeek/metrics/NHD.xsl [hunk 1]
```diff
@@ -0,0 +1,89 @@
+<?xml version="1.0"?>
+<!--
+The MIT License (MIT)
+
+Copyright (c) 2017 Yegor Bugayenko
```
Comment (amihaiemil) [inline src/main/resources/org/jpeek/metrics/NHD.xsl]: @llorllale 2017-2018 here

File: src/main/resources/org/jpeek/metrics/NHD.xsl [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+SOFTWARE.
+-->
+<xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform" version="2.0">
+  <xsl:param name="ctors" select="0"/>
+  <xsl:template match="skeleton">
+    <metric>
+      <xsl:apply-templates select="@*"/>
+      <title>NHD</title>
+      <description>
+        <xsl:text>The NHD (Normalized Hamming Distance) class cohesion metric
```
Comment (amihaiemil) [inline src/main/resources/org/jpeek/metrics/NHD.xsl]: @llorllale Can you put this text also in the PRs description?

File: src/test/java/org/jpeek/MetricsTest.java [hunk 3]
```diff
@@ -35,6 +35,12 @@
...
 import org.junit.runners.Parameterized;
 
+// @todo #93:30min NHD needs to be tested against the following after #103 is
+//  fixed: NoMethods, OneVoidMethodWithoutParams, WithoutAttributes,
+//  OneMethodCreatesLambda. NHD score for all these is "NaN".
+// @todo #93:30min NHD calculation needs to take into account the method's
+//  visibility, which should be configurable and implemented after #101 is
+//  fixed.
 /**
  * Tests for all metrics.
```
Comment (amihaiemil) [inline src/test/java/org/jpeek/MetricsTest.java]: @llorllale The 2 puzzles from above can be in this JavaDoc, Starting at this line.

File: src/test/java/org/jpeek/MetricsTest.java [hunk 4]
```diff
@@ -35,6 +35,12 @@
 import org.junit.runner.RunWith;
 import org.junit.runners.Parameterized;
 
+// @todo #93:30min NHD needs to be tested against the following after #103 is
```
Comment (amihaiemil) [inline src/test/java/org/jpeek/MetricsTest.java]: @llorllale As I understand from #103, the overall report will fail if one of the metrics returns ``NaN``, right? So, until 103 is fixed, wouldn't it be better if this metric returned ``0`` instead of ``NaN``? Then, when 103 is fixed we go back to ``NaN`` here and add those tests.

If so, please update the puzzle to reflect also that.

Comment (llorllale) [reply src/test/java/org/jpeek/MetricsTest.java]: @amihaiemil from #103, it's the way tests are structured in `MetricsTest` that is not compatible with reports that may return `NaN` as a value. But for the regular jpeek use, the report itself does not "fail". In other words, the report generation does not throw errors and the output html renders without issues in a browser. It *could* be that the mean, average, variance, and so on, are thrown off for NHD if any classes result in `NaN`, but having the implementation return `0` will probably have the same effect.

So in other words, I believe we gain nothing by the change you suggested. WDYT?

File: src/test/java/org/jpeek/MetricsTest.java (hunk 3 above)
Comment (llorllale) [reply src/test/java/org/jpeek/MetricsTest.java]: @amihaiemil this is tricky because qulice starts to complain a lot. Starting with `First sentence in a comment should start with a capital letter (RegexpMultilineCheck)`. I'd need to disable some checks on this file; conversely, someone needs to remember to remove the checks when they complete the puzzle.

Do you feel strongly about having them in the javadoc?

File: src/main/resources/org/jpeek/metrics/NHD.xsl (hunk 1 above)
Comment (llorllale) [reply src/main/resources/org/jpeek/metrics/NHD.xsl]: @amihaiemil fixed

File: src/main/resources/org/jpeek/metrics/NHD.xsl (hunk 2 above)
Comment (llorllale) [reply src/main/resources/org/jpeek/metrics/NHD.xsl]: @amihaiemil done

File: src/test/java/org/jpeek/MetricsTest.java (hunk 4 above)
Comment (amihaiemil) [reply src/test/java/org/jpeek/MetricsTest.java]: @llorllale I see. Let's leave it like that then. I figured that if the tests wouldn't work, then the App wouldn't as well :)

File: src/test/java/org/jpeek/MetricsTest.java (hunk 3 above)
Comment (amihaiemil) [reply src/test/java/org/jpeek/MetricsTest.java]: @llorllale I see, ok

**Comment by amihaiemil** [comment]
@rultor good to merge

**Comment by yegor256** [comment]
@rultor merge

**Comment by amihaiemil** [comment]
@yegor256 How long is it supposed to pass until 0crat notices a merged PR? It's been 20hrs already, should I file an issue?

**Comment by yegor256** [comment]
@0crat status

**Comment by 0crat** [comment]
There is an unrecoverable failure on my side. Please, submit it [here](https://github.com/zerocracy/farm/issues):

```
java.lang.IllegalStateException: thread 'BkParallel-8' in group 'main' interrupted
	at com.jcabi.aspects.aj.MethodLogger.wrap(MethodLogger.java:184)
	at com.jcabi.aspects.aj.MethodLogger.ajc$inlineAccessMethod$com_jcabi_aspects_aj_MethodLogger$com_jcabi_aspects_aj_MethodLogger$wrap(MethodLogger.java:1)
	at com.jcabi.aspects.aj.MethodLogger.wrapClass(MethodLogger.java:140)
	at com.jcabi.github.Comment$Smart.issue(Comment.java:181)
	at com.zerocracy.radars.github.ReSafe.lambda$react$1(ReSafe.java:70)
	at org.cactoos.func.FuncOf.lambda$new$3(FuncOf.java:88)
	at org.cactoos.func.FuncOf.apply(FuncOf.java:104)
	at org.cactoos.func.FuncWithFallback.apply(FuncWithFallback.java:90)
	at org.cactoos.func.IoCheckedFunc.lambda$apply$0(IoCheckedFunc.java:60)
	at org.cactoos.func.IoCheckedBiFunc.apply(IoCheckedBiFunc.java:66)
	at org.cactoos.func.IoCheckedFunc.apply(IoCheckedFunc.java:61)
	at com.zerocracy.radars.github.ReSafe.react(ReSafe.java:80)
	at com.zerocracy.radars.github.ReOnComment.send(...
```

[1.0-SNAPSHOT](https://github.com/zerocracy/farm/tree/1.0-SNAPSHOT) /cc @yegor256

**Comment by yegor256** [comment]
@0crat status

**Comment by yegor256** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/jpeek/issues/113#issuecomment-361231410))

@yegor256 This is what I know about this job, as in [§32](http://datum.zerocracy.com/pages/policy.html#32):

  * The job [#113](https://github.com/yegor256/jpeek/issues/113) is [in scope](http://datum.zerocracy.com/pages/policy.html#14) for 3days
  * The role is `REV`
  * The job is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil) for [3days](http://datum.zerocracy.com/pages/policy.html#8)
  * There is a monetary reward attached
  * The job doesn't have any [impediments](http://datum.zerocracy.com/pages/policy.html#9)
  * The [budget](http://datum.zerocracy.com/pages/policy.html#4) is 15 minutes
  * These users are banned and won't be assigned:
    * @llorllale[/z](https://www.0crat.com/u/llorllale): This user reported the ticket

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/jpeek/issues/113#issuecomment-361232509))

@yegor256 This is what I know about this job, as in [§32](http://datum.zerocracy.com/pages/policy.html#32):

  * The job [#113](https://github.com/yegor256/jpeek/issues/113) is [in scope](http://datum.zerocracy.com/pages/policy.html#14) for 3days
  * The role is `REV`
  * The job is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil) for [3days](http://datum.zerocracy.com/pages/policy.html#8)
  * There is a monetary reward attached
  * The job doesn't have any [impediments](http://datum.zerocracy.com/pages/policy.html#9)
  * The [budget](http://datum.zerocracy.com/pages/policy.html#4) is 15 minutes
  * These users are banned and won't be assigned:
    * @llorllale[/z](https://www.0crat.com/u/llorllale): This user reported the ticket

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/jpeek/issues/113#issuecomment-361232509))

@yegor256 This is what I know about this job, as in [§32](http://datum.zerocracy.com/pages/policy.html#32):

  * The job [#113](https://github.com/yegor256/jpeek/issues/113) is [in scope](http://datum.zerocracy.com/pages/policy.html#14) for 3days
  * The role is `REV`
  * The job is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil) for [3days](http://datum.zerocracy.com/pages/policy.html#8)
  * There is a monetary reward attached
  * The job doesn't have any [impediments](http://datum.zerocracy.com/pages/policy.html#9)
  * The [budget](http://datum.zerocracy.com/pages/policy.html#4) is 15 minutes
  * These users are banned and won't be assigned:
    * @llorllale[/z](https://www.0crat.com/u/llorllale): This user reported the ticket

**Comment by yegor256** [comment]
@amihaiemil let's wait a bit

**Comment by amihaiemil** [comment]
@yegor256 sure

**Comment by 0crat** [comment]
Order was successfully finished: +15 points just awarded to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), total is [+180](https://www.0crat.com/u/amihaiemil)

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://datum.zerocracy.com/pages/policy.html#28): +15 points just awarded to @yegor256[/z](https://www.0crat.com/u/yegor256), total is [+8655](https://www.0crat.com/u/yegor256)

**Comment by 0crat** [comment]
Job `gh:yegor256/jpeek#113` is not assigned, can't get start time

**Comment by 0crat** [comment]
The job [#113](https://github.com/yegor256/jpeek/issues/113) is now out of scope

**Comment by 0crat** [comment]
Job [#113](https://github.com/yegor256/jpeek/issues/113) was not in scope
